### PR TITLE
#673 fix string resources generation for jvm and apple targets

### PR DESCRIPTION
**Pull request by ExNDY**
Added text with newline in sample

### CODE REVIEW

File: resources-generator/src/main/kotlin/dev/icerock/gradle/utils/StringExt.kt [hunk 1]
```diff
@@ -42,11 +41,7 @@ internal val String.flatName: String
     get() = this.remove('.')
 
 internal fun String.convertXmlStringToLocalizationValue(): String {
-    return StringEscapeUtils.unescapeXml(this).let {
-        UnicodeUnescaper().translate(
-            StringEscapeUtils.escapeJava(it)
-        )
-    }
+    return StringEscapeUtils.unescapeXml(this)
```
Comment (Alex009) [inline resources-generator/src/main/kotlin/dev/icerock/gradle/utils/StringExt.kt:44]: Are you sure that it works correctly?
Bug was about iOS.

Add please cases with " ' % @ * {} symbols too

Comment (ExNDY) [reply resources-generator/src/main/kotlin/dev/icerock/gradle/utils/StringExt.kt:44]: i'm add sample text with symbols, for " and ' need mirroring for correct work on all platforms (exclude js: that platform incorrect work with mirroring, my opinion)

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 package dev.icerock.gradle.generator.platform.js
 
-private val messageFormatRegex = "%(.)(?:\\\$(.))?".toRegex()
+private val messageFormatRegex = "(%[a-z])|(%[.][a-z])".toRegex()
```
Comment (Alex009) [inline resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt:7]: need to fix positional arguments and add tests
```
assertEquals(
    expect = "some string {0}",
    actual = "some string %s".convertToMessageFormat()
)
```

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/plural/PluralResourceGenerator.kt [hunk 3]
```diff
@@ -77,8 +77,9 @@ internal class PluralResourceGenerator(
 
                 val quantity: String = item.attributes.getNamedItem("quantity").textContent.trim()
                 val value: String = item.textContent
+                val strictValue: String = if (strictLineBreaks) value else value.removeLineWraps()
 
-                pluralMap[quantity] = if (strictLineBreaks) value else value.removeLineWraps()
+                pluralMap[quantity] = strictValue.replace("\\n", "\n")
```
Comment (Alex009) [inline resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/plural/PluralResourceGenerator.kt:82]: remove duplication of this part.
```
pluralMap[quantity] = value.processXmlTextContent(strictLineBreaks)
```

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/string/StringResourceGenerator.kt [hunk 4]
```diff
@@ -66,8 +66,9 @@ internal class StringResourceGenerator(
             val stringNode: Node = stringNodes.item(i)
             val name: String = stringNode.attributes.getNamedItem("name").textContent
             val value: String = stringNode.textContent
+            val strictValue: String = if (strictLineBreaks) value else value.removeLineWraps()
 
-            resultMap[name] = if (strictLineBreaks) value else value.removeLineWraps()
+            resultMap[name] = strictValue.replace("\\n", "\n")
```
Comment (Alex009) [inline resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/string/StringResourceGenerator.kt:71]: remove duplication of this part.
```
resultMap[name] = value.processXmlTextContent(strictLineBreaks)
```

**Review by Alex009 — CHANGES_REQUESTED** [review]

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt [hunk 5]
```diff
@@ -4,7 +4,10 @@
 
 package dev.icerock.gradle.generator.platform.js
 
-private val messageFormatRegex = "%(.)(?:\\\$(.))?".toRegex()
+import dev.icerock.gradle.utils.remove
+
+private val messageFormatRegex =
+    "%(\\d+\\\$)?(\\d*\\.?\\d+)?[b,B,h,H,s,S,c,C,d,o,x,X,e,E,f,g,G,a,A,t,T]+".toRegex()
```
Comment (Alex009) [inline resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt:10]: why `\\.?` ? is format "%2d" is ok?

File: resources-generator/src/test/kotlin/dev/icerock/gradle/generator/jsGenerator/JsMessageFormatTest.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        )
+    }
+
+    @Test
+    fun textWithSingleWithDecimalArgumentTest() {
+        assertEquals(
+            expected = "Only {0} percents of people like animals",
+            actual = "Only %.2d percents of people like animals".convertToMessageFormat()
+        )
+    }
```
Comment (Alex009) [inline resources-generator/src/test/kotlin/dev/icerock/gradle/generator/jsGenerator/JsMessageFormatTest.kt:50]: please add test with multiple not positional arguments

**Review by Alex009 — CHANGES_REQUESTED** [review]

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt (hunk 5 above)
Comment (ExNDY) [reply resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt:10]: %2f is correct for format

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/plural/PluralResourceGenerator.kt (hunk 3 above)
Comment (ExNDY) [reply resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/plural/PluralResourceGenerator.kt]: done

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/string/StringResourceGenerator.kt (hunk 4 above)
Comment (ExNDY) [reply resources-generator/src/main/kotlin/dev/icerock/gradle/generator/resources/string/StringResourceGenerator.kt]: done

File: resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt (hunk 5 above)
Comment (ExNDY) [reply resources-generator/src/main/kotlin/dev/icerock/gradle/generator/platform/js/JsParamsHelper.kt]: removed , from regex

File: resources-generator/src/test/kotlin/dev/icerock/gradle/generator/jsGenerator/JsMessageFormatTest.kt (hunk 6 above)
Comment (ExNDY) [reply resources-generator/src/test/kotlin/dev/icerock/gradle/generator/jsGenerator/JsMessageFormatTest.kt:50]: added

**Review by Alex009 — APPROVED** [review]

**Pull request merged**